### PR TITLE
logging: add syst format support for existing logging system

### LINKED ISSUE #19841: MIPI Sys-T logging/tracing support
Zephyr shall support Sys-T trace via the existing logging subsystem as a custom format that can use existing logging backends (
serial for example) to send logging data in sys-t format to a host. This feature will use existing sys-t library (https://github.com/MIPI-Alliance/public-mipi-sys-t).

We need a module in Zephyr to be included by west when enabling this feature in Zephyr. This will be based on https://github.com/MIPI-Alliance/public-mipi-sys-t

### PR DESCRIPTION
**Pull request by wentongwu**
add syst format support for existing logging system, the output of sys-t format can be decoded by the existing decoder.

Fixes: #19841.

### CODE REVIEW

File: west.yml [hunk 1]
```diff
@@ -21,6 +21,8 @@ manifest:
   remotes:
     - name: upstream
       url-base: https://github.com/zephyrproject-rtos
+    - name: syst_upstream
```
Comment (andrewboie) [inline west.yml]: Hmm shouldn't we be mirroring this at the project level, instead of adding a new upstream?

Comment (nashif) [reply west.yml]: this is still WIP I guess :)

Comment (wentongwu) [reply west.yml]: @andrewboie @nashif put the changes for syst on my own fork first, when everything is good, will do the mirroring at the project level.

**Comment by wentongwu** [comment]
the output of failed test cases is like below which is what we wanted and also it can be decoded by the decoder, the CI failure may be caused by the wrong configure of the added sample, I will fix it.
 
```
SYS-T RAW DATA: 0200000B2A2A2A2A2A20426F6F74696E67205A6570687972204F53206275696C64207A65706879722D76322E302E302D313038382D67303264333035623465323134202A2A2A2A2A0A00
SYS-T RAW DATA: 2200000B4572726F72206D657373616765206578616D706C652E00
SYS-T RAW DATA: 3200000B5761726E696E67206D657373616765206578616D706C652E00
SYS-T RAW DATA: 4200000B496E666F206D657373616765206578616D706C652E00
SYS-T RAW DATA: 7200000B4465627567206D657373616765206578616D706C652E00
SYS-T RAW DATA: 7200000B4465627567206D657373616765206578616D706C652E00
SYS-T RAW DATA: 7200000B4465627567206D657373616765206578616D706C652C2025640001000000
SYS-T RAW DATA: 7200000B4465627567206D657373616765206578616D706C652C2025642C202564000100000002000000
SYS-T RAW DATA: 7200000B4465627567206D657373616765206578616D706C652C2025642C2025642C20256400010000000200000003000000
SYS-T RAW DATA: 7200000B4465627567206D657373616765206578616D706C652C2025642C2025642C2025642C2030782532780001000000020000000300000004000000
SYS-T RAW DATA: 2600001A03D2040000080102030405060708
SYS-T RAW DATA: 3600001A03D2040000080102030405060708
SYS-T RAW DATA: 4600001A03D2040000080102030405060708
SYS-T RAW DATA: 7600001A03D2040000080102030405060708
SYS-T RAW DATA: 0200000B68656C6C6F207379732D74206F6E20626F617264206D7073325F616E33383500
```

**Comment by wentongwu** [comment]
@nashif could you please give some comments for this PR? Thanks

File: west.yml [hunk 2]
```diff
@@ -100,6 +102,10 @@ manifest:
     - name: littlefs
       path: modules/fs/littlefs
       revision: fe9572dd5a9fcf93a249daa4233012692bd2881d
+    - name: public-mipi-sys-t
```
Comment (nashif) [inline west.yml]: please use github.com:zephyrproject-rtos/mipi-sys-t.git instead and put this in its own commit.

File: include/mipi_syst/platform.h [hunk 3]
```diff
@@ -0,0 +1,106 @@
+/*
```
Comment (nashif) [inline include/mipi_syst/platform.h]: this is not a zephyr public API, you should move it under subsystem/logging or subsystem/debug.

File: modules/Kconfig.syst [hunk 4]
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2019 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+config MIPI_SYST_LIB
+	bool "MIPI SyS-T Library Support"
+	select NEWLIB_LIBC
```
Comment (nashif) [inline modules/Kconfig.syst]: why do you need newlib?
Also, move this to the same commit where you introduce the module in west.yml.

File: subsys/logging/log_output_syst.c [hunk 5]
```diff
@@ -24,8 +24,8 @@
```
Comment (nashif) [inline subsys/logging/log_output_syst.c]: you probably want to fold this commit into the previous commit introducing this file.

File: samples/subsys/logging/syst/src/main.c [hunk 6]
```diff
@@ -14,7 +14,7 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME, LOG_LEVEL_DBG);
```
Comment (nashif) [inline samples/subsys/logging/syst/src/main.c]: squash into original commit introducing the change please

File: samples/subsys/logging/syst/sample.yaml [hunk 7]
```diff
@@ -4,3 +4,8 @@ tests:
   sample.subsys.logging.syst:
     platform_whitelist: mps2_an385
     tags: logging
+    harness: console
```
Comment (nashif) [inline samples/subsys/logging/syst/sample.yaml]: squash into original commit please

File: subsys/logging/CMakeLists.txt [hunk 8]
```diff
@@ -48,6 +47,16 @@ if(NOT CONFIG_LOG_MINIMAL)
     CONFIG_LOG_BACKEND_QEMU_X86_64
     log_backend_qemu_x86_64.c
   )
+
+  zephyr_sources_ifdef(
```
Comment (nashif) [inline subsys/logging/CMakeLists.txt]: That does not seems like the right approach here, @nordic-krch  can you please provide some feedback on how to support a different output back-end?

**Review by nashif — DISMISSED** [review]

File: subsys/logging/CMakeLists.txt (hunk 8 above)
Comment (wentongwu) [reply subsys/logging/CMakeLists.txt]: @nashif IMHO back-end doesn't take care the output format, just pushing the data out with his own protocol(uart, usb, etc.). it's the place where in log_output.c logging subsystem decide the output format, so I replace the original one with sys-t format. I don't do the custom in the front-end, because I don't want to sys-t feature require more stack from the thread which use the front-end Marcos(LOG_INFO, LOG_ERR, etc).

@nordic-krch could you please help give your comments? Thanks.

File: modules/Kconfig.syst (hunk 4 above)
Comment (wentongwu) [reply modules/Kconfig.syst]: sys-t lib requires some header files which can't be founded in Minimal libc.

File: west.yml (hunk 2 above)
Comment (wentongwu) [reply west.yml]: ok

File: include/mipi_syst/platform.h (hunk 3 above)
Comment (wentongwu) [reply include/mipi_syst/platform.h]: ok, Thanks

**Comment by wentongwu** [comment]
@nashif @nordic-krch merge window will close in 2 or 3 weeks, really appreciate you can give further comments about this PR. Thanks.

**Comment by nordic-krch** [comment]
@wentongwu i don't like forking `log_output.c` into `log_output_syst.c`. there is a log of code redundancy between those two and it forces syst format on all backends. Isn't it possible to extend log_output will additional flags? Something similar to `LOG_OUTPUT_FLAG_FORMAT_SYSLOG`. This will allow to use syst format only for specific backend.

File: modules/Kconfig.syst [hunk 9]
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2019 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+config MIPI_SYST_LIB
```
Comment (nordic-krch) [inline modules/Kconfig.syst]: I would add option to the logger kconfig (like `LOG_MIPI_SYST_ENABLE`) which select this option. Otherwise there is no explicit connection between the logger and this library.

Comment (wentongwu) [reply modules/Kconfig.syst]: Thanks,

File: subsys/logging/log_backend_uart.c [hunk 10]
```diff
@@ -31,7 +31,12 @@ LOG_OUTPUT_DEFINE(log_output, char_out, &buf, 1);
 static void put(const struct log_backend *const backend,
 		struct log_msg *msg)
 {
-	log_backend_std_put(&log_output, 0, msg);
+	u32_t flag = 0;
+
+#ifdef CONFIG_LOG_BACKEND_UART_SYST_ENABLE
```
Comment (nordic-krch) [inline subsys/logging/log_backend_uart.c]: you could use `IS_ENABLED()`
something like
`u32_t flag = IS_ENABLED(SYST_ENABLE) ? LOG_OUTPUT_FLAG_FORMAT_SYST : 0;`
would look more concise, imo.

File: subsys/logging/log_output.c [hunk 11]
```diff
@@ -506,6 +518,13 @@ void log_output_msg_process(const struct log_output *log_output,
 	bool raw_string = (level == LOG_LEVEL_INTERNAL_RAW_STRING);
 	int prefix_offset;
 
+#ifdef CONFIG_LOG_MIPI_SYST_ENABLE
+	if (flags & LOG_OUTPUT_FLAG_FORMAT_SYST) {
```
Comment (nordic-krch) [inline subsys/logging/log_output.c]: i cases like that (runtime + compile time condition) i prefer to create local function (e.g. `is_format_syst_enabled(flags)`) which simply wraps `IS_ENABLED(SYST_ENABLE) && (flags & SYST_FLAG)`. that reduces ifdefs to 0 (no need to ifdef function declarations too)

File: subsys/logging/log_backend_uart.c (hunk 10 above)
Comment (wentongwu) [reply subsys/logging/log_backend_uart.c]: Thanks, understand the IS_ENABLED

File: subsys/logging/log_output.c (hunk 11 above)
Comment (wentongwu) [reply subsys/logging/log_output.c]: Thanks

File: west.yml (hunk 2 above)
Comment (wentongwu) [reply west.yml]: @nashif have already used zephyrproject-rtos/mipi-sys-t.git.

**Comment by wentongwu** [comment]
> @wentongwu i don't like forking `log_output.c` into `log_output_syst.c`. there is a log of code redundancy between those two and it forces syst format on all backends. Isn't it possible to extend log_output will additional flags? Something similar to `LOG_OUTPUT_FLAG_FORMAT_SYSLOG`. This will allow to use syst format only for specific backend.

The initial idea is that I don't want to break the exiting log_output.c, but it needs the existing interfaces to have logging system work well, so I forked log_output.c into log_output_syst.c which implemented the needed APIs.

Now I restructure the code, put sys-t related code from log_output_syst.c to https://github.com/zephyrproject-rtos/mipi-sys-t and only leave the APIs in log_output_syst.c needed by log_output.c if enabled sys-t format for that back-end. Thanks for your suggestions, could you please take another look the new code? Thanks

File: samples/subsys/logging/syst/src/main.c [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+	};
+
+
+	/* standard print */
+	LOG_ERR("Error message example.");
+	LOG_WRN("Warning message example.");
+	LOG_INF("Info message example.");
+	LOG_DBG("Debug message example.");
+
+	for (int i = 0; i <= 4; i++) {
```
Comment (nordic-krch) [inline samples/subsys/logging/syst/src/main.c]: what's the purpose of this for loop? why not simply put 5 log messages one by one?

Comment (andrewboie) [reply samples/subsys/logging/syst/src/main.c]: this confuses me too

File: subsys/logging/log_backend_rtt.c [hunk 13]
```diff
@@ -221,7 +221,10 @@ LOG_OUTPUT_DEFINE(log_output, IS_ENABLED(CONFIG_LOG_BACKEND_RTT_MODE_BLOCK) ?
 static void put(const struct log_backend *const backend,
 		struct log_msg *msg)
 {
-	log_backend_std_put(&log_output, 0, msg);
+	u32_t flag = IS_ENABLED(CONFIG_LOG_BACKEND_RTT_SYST_ENABLE) ?
```
Comment (andrewboie) [inline subsys/logging/log_backend_rtt.c]: it doesn't seem good that we have to patch all the backends like this specifically for one output formatting type. it doesn't seem like this will scale well to additional format types as they are added. is there some other way we can do this in more centralized code?

File: samples/subsys/logging/syst/src/main.c (hunk 12 above)
Comment (wentongwu) [reply samples/subsys/logging/syst/src/main.c]: my fault

File: subsys/logging/log_backend_rtt.c (hunk 13 above)
Comment (nordic-krch) [reply subsys/logging/log_backend_rtt.c]: alternative is enabling syst formatting for all backends at once. However, as i'm not familiar with syst i don't know what is the exact use case. How does it match with formatting used for net or bluetooth.

Comment (wentongwu) [reply subsys/logging/log_backend_rtt.c]: currently I don't think there is a centralized way to enable syst formatting for all backends, because besides standard logger backends, there are non-standard backends like net backend or bluetooth backend.

Comment (wentongwu) [reply subsys/logging/log_backend_rtt.c]: @nordic-krch @andrewboie @nashif besides centralized issue, is there any other concerns for this PR? Thanks

File: subsys/logging/log_output_syst.c [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+		break;
+	}
+}
+
+static void raw_string_print(struct log_msg *msg,
+			const struct log_output *log_output)
+{
+	char buf[CONFIG_LOG_STRDUP_MAX_STRING + 1];
+	size_t length = CONFIG_LOG_STRDUP_MAX_STRING;
+	u32_t severity = level_to_syst_severity(log_msg_level_get(msg));
```
Comment (nordic-krch) [inline subsys/logging/log_output_syst.c]: raw string has custom `LOG_LEVEL_INTERNAL_RAW_STRING` (5) here it will be converted to debug level. Not sure if that is an expected result. Raw string is used when printk goes through logger. Alternatively severity can be fixed here to 0.

Comment (wentongwu) [reply subsys/logging/log_output_syst.c]: @nordic-krch yes, we use raw string if printk goes through logger, and also I have it tested in sample code like below.
```

        /* raw string */
        printk("hello sys-t on board %s\n", CONFIG_BOARD);
```
And checked code again, it seems `LOG_LEVEL_INTERNAL_RAW_STRING` already equals 0 as below marco defined. If anything wrong, please correct me. Thanks
`#define LOG_LEVEL_INTERNAL_RAW_STRING LOG_LEVEL_NONE`

Comment (nordic-krch) [reply subsys/logging/log_output_syst.c]: my bad, somehow i thought it had different number.

**Comment by nordic-krch** [comment]
PR looks good. I think it would be good to extend conceptual documentation to mention this feature Frankly speaking i'm not exactly sure how syst is going to be used.

**Review by nordic-krch — APPROVED** [review]
Hoping to see conceptual documentation extension in another PR.

File: subsys/logging/Kconfig [hunk 15]
```diff
@@ -492,6 +492,12 @@ config LOG_BACKEND_NET_MAX_BUF_SIZE
 	  IPv6 the size is 1180 octets. As each buffer will use RAM, the value
 	  should be selected so that typical messages will fit the buffer.
 
+config LOG_BACKEND_NET_SYST_ENABLE
```
Comment (jukkar) [inline subsys/logging/Kconfig]: I was not marked as a reviewer for this PR so did not had a chance to comment this. I think this was merged a bit prematurely as now if user enables `CONFIG_LOG_BACKEND_NET` and `CONFIG_LOG_BACKEND_NET_SYST_ENABLE`, the loggers net backend prepares to send syslog formatted messages out but this will fail unless user sets the syslog server hostname. The code is prepared to handle that issue but there is now just useless code in the system if user only meant to activate syst logging and not syslog one.
Better would have been to have separate option for enabling syslog formatted network logging, just like the syst one.

**Comment by wentongwu** [comment]
> Hoping to see conceptual documentation extension in another PR.

I will, thanks.
